### PR TITLE
New rule category `XP`, implement `not-array-agnostic-numpy` (`XP001`)

### PR DESCRIPTION
**Pull request by lucascolley**
## Summary

Closes gh-8615

- New rule category `XP`
- New rule `NotArrayAgnosticNumPy`

Aiming for violations for all non-standard NumPy functions and methods and uses of them. Some of these will have standard aliases or alternatives. We should try to catch as many of the 'compatible' and 'strictness' changes of https://numpy.org/doc/stable/reference/array_api.html as possible (the 'breaking' changes should be caught separately - by `NP201` or perhaps `NP202` is also needed).

## Test Plan

_WIP_

### CODE REVIEW

File: crates/ruff_linter/src/registry.rs [hunk 1]
```diff
@@ -208,6 +208,9 @@ pub enum Linter {
     /// Ruff-specific rules
     #[prefix = "RUF"]
     Ruff,
+    /// [Array-agnosticism](https://data-apis.org/array-api/latest/)
```
Comment (lucascolley) [inline crates/ruff_linter/src/registry.rs:211]:
```suggestion
    /// [Array-agnosticism](https://github.com/data-apis/array-api)
```

File: crates/ruff_linter/src/rules/xp/rules/not_array_agnostic_numpy.rs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+                },
+            }),
+            ["numpy", func] if ARRAY_API_FUNCTIONS.contains(func) => None,
+            ["numpy", func] => Some(Replacement {
+                existing: func,
+                details: Details::Manual {
+                    guideline: Some(concat!(
+                        "xp.",
+                        stringify!($func),
+                        "is not in the array API standard"
```
Comment (lucascolley) [inline crates/ruff_linter/src/rules/xp/rules/not_array_agnostic_numpy.rs:374]:
```suggestion
                        " is not in the array API standard"
```

**Comment by MichaReiser** [comment]
Thank you @lucascolley for this rule proposal and implementation. I really like the idea of it. I think such a rule would be very helpful for our users. Unfortunately, Ruff doesn't have a good concept today for very opinionated rules or rules that intentionally restrict the allowed APIs. Ruff also lacks a good default rule set, and, unfortunately, many users assume that `ALL` is our recommended rule set. That means users using `--select ALL` would automatically opt into this new restricting rule, and they do not necessarily have the experience to judge whether the pattern flagged by this rule is indeed a "problem" for them or if they are better off by ignoring the rule. We want to hold off from merging restricting rules until #1774 is complete. We're actively working on it, and we hope to have a good place for rules like the one you proposed here. 

I'm sorry that it took us so long to give that feedback to you, and I hope we can get this rule landed once #1774 is completed. Thanks again.

**Comment by lucascolley** [comment]
Thanks for the reply @MichaReiser , no worries! Happy to return to this once gh-1774 is resolved.

> That means users using --select ALL would automatically opt into this new restricting rule

Does that include `preview` rules? If so, should it not?

**Comment by MichaReiser** [comment]
> Does that include preview rules? If so, should it not?

It does include preview rules when running ruff with `preview = true` (you have to opt into preview). We considered accepting rules as preview only but we decided against it because accepting rules without a clear path to stabilization.

**Pull request closed without merging**